File: orm.py
from abc import ABC, abstractmethod
import sqlite3

class NOTNULL:
    def __class_getitem__(cls, item):
        obj = cls()
        obj.base_type = item
        return obj

    def __str__(self):
        return f"{type_map.get(self.base_type, 'TEXT')} NOT NULL"


class PRIMARYKEY:
    def __class_getitem__(cls, item):
        obj = cls()
        obj.base_type = item
        return obj

    def __str__(self):
        return f"{type_map[self.base_type]} PRIMARY KEY AUTOINCREMENT"

class FOREIGNKEY:
    def __class_getitem__(cls, target_model):
        # target_model is the *actual* class (like PlayList)
        obj = cls()
        obj.target_model = target_model     # store the model class
        return obj

    def __str__(self):
        # This is used during table creation
        table_name = self.target_model.TableName()
        return f"INTEGER REFERENCES {table_name}(id)"

class Unique:
    def __class_getitem__(cls, item):
        obj = cls()
        obj.base_type = item
        return obj

    def __str__(self):
        return f"{type_map[self.base_type]} UNIQUE"
    

class DEFAULT:
    def __init__(self, value):
        self.value = value

    def __str__(self):
        if isinstance(self.value, str):
            return f"DEFAULT '{self.value}'"
        return f"DEFAULT {self.value}"

class CHECK:
    def __init__(self, expr):
        self.expr = expr

    def __str__(self):
        return f"CHECK({self.expr})"



class Condition:
    def __init__(self, expr):
        self.expr = expr
    

    def __and__(self, other: 'Condition'):
        return Condition(f"{self.expr} AND {other.expr}")

    def __or__(self, other: 'Condition'):
        return Condition(f"{self.expr} OR {other.expr}")

    def __invert__(self):
        return Condition(f"NOT {self.expr}")

    def __str__(self):
        return self.expr

class Field:
    def __init__(self, name, model=None):
        self.name = name
        self.model = model  # <-- NEW: who owns this field

    def __eq__(self, value):
        return Condition(f"{self.full_name()} = {repr(value)}")

    def __ne__(self, value):
        return Condition(f"{self.full_name()} != {repr(value)}")

    def __gt__(self, value):
        return Condition(f"{self.full_name()} > {repr(value)}")

    def __lt__(self, value):
        return Condition(f"{self.full_name()} < {repr(value)}")

    def __ge__(self, value):
        return Condition(f"{self.full_name()} >= {repr(value)}")

    def __le__(self, value):
        return Condition(f"{self.full_name()} <= {repr(value)}")

    def in_(self, values):
        vals = ", ".join(repr(v) for v in values)
        return Condition(f"{self.full_name()} IN ({vals})")

    def like(self, pattern):
        return Condition(f"{self.full_name()} LIKE {repr(pattern)}")

    def full_name(self):
        if self.model:
            return f"{self.model.TableName()}.{self.name}"
        return self.name

type_map = {
    int: "INTEGER",
    str: "TEXT",
    float: "REAL",
    bool: "INTEGER"
}



class ModelMeta(type):
    def __new__(mcls, name, bases, attrs):
        cls = super().__new__(mcls, name, bases, attrs)
        if name == "BaseModel":
            return cls

        cls._foreign_keys = {}

        for field_name, field_type in getattr(cls, "__annotations__", {}).items():
            if isinstance(field_type, FOREIGNKEY):
                cls._foreign_keys[field_type.target_model] = field_name 
            setattr(cls, field_name, Field(field_name,model=cls))

        return cls
    

class BaseModel(metaclass=ModelMeta):
    DB_PATH = 'app.db'
    _connection = None

    def __init__(self, **kwargs):
        for name in self.fields():
            setattr(self, name, kwargs.get(name))

        def __init_subclass__(cls):
            super().__init_subclass__()
            for field_name in cls.fields().keys():
                setattr(cls, field_name, Field(field_name))

    @classmethod 
    def connection(cls):
        if cls._connection is None:
            cls._connection = sqlite3.connect(cls.DB_PATH)
        return cls._connection

    @classmethod
    def constraint(cls):
        return cls.CONSTRAINT

    @classmethod
    def all(cls):
        with cls.connection() as conn:
            cur = conn.cursor()
            cur.execute(f"SELECT * FROM {cls.TableName()}")
            rows = cur.fetchall()
        
        field_name = list(cls.fields().keys())
        objects = []
        for row in rows:
            data = {name: value for name, value in zip(field_name, row)}
            obj = cls(**data)
            objects.append(obj)
        return objects


    @classmethod
    def update(cls, where: dict = None, **kwargs):
        """
        UPDATE {table_name}
        SET k = v {from kwargs}
        WHERE {conditions from where dict}
        """

        if not kwargs:
            raise ValueError("No fields to update")

        set_clause = ", ".join([f"{k} = ?" for k in kwargs.keys()])
        set_values = list(kwargs.values())

        where_clause = ""
        where_values = []
        if where:
            where_clause = " AND ".join([f"{k} = ?" for k in where.keys()])
            where_values = list(where.values())

        sql = f"UPDATE {cls.TableName()} SET {set_clause}"
        if where_clause:
            sql += f" WHERE {where_clause}"


        conn = cls.connection()
        cur = conn.cursor()
        cur.execute(sql,set_values + where_values)
        conn.commit()



    @classmethod
    def filter(cls, *args, **kwargs):
        joins = []
        where_clause = ""

        if args and isinstance(args[0], Condition):
            where_clause = str(args[0])
            for model in BaseModel.__subclasses__():
                if model.TableName() in where_clause and model != cls:
                    joins.append(model)

        else:
            conditions = []
            for k, v in kwargs.items():
                val = f"'{v}'" if isinstance(v, str) else v
                conditions.append(f"{cls.TableName()}.{k} = {val}")
            where_clause = " AND ".join(conditions)

        sql = f"SELECT {cls.TableName()}.* FROM {cls.TableName()}"

        # Add JOINs automatically
        for j in joins:
            fk_field = cls._foreign_keys[j]

            if fk_field:
                sql += f" JOIN {j.TableName()} ON {cls.TableName()}.{fk_field} = {j.TableName()}.id"
                
        sql += f" WHERE {where_clause}"

        conn = cls.connection()
        cur = conn.cursor()
        cur.execute(sql)
        rows = cur.fetchall()

        field_name = list(cls.fields().keys())
        objects = []
        for row in rows:
            data = {name: value for name, value in zip(field_name, row)}
            obj = cls(**data)
            objects.append(obj)
        return objects

        
    @classmethod
    def primary_key(cls) -> str:
        return getattr(cls,"PK", "id")

    @classmethod    
    @abstractmethod
    def TableName(cls) -> str:
        return cls.__name__.lower()


    @classmethod
    def fields(cls) -> dict[str,type]:
        return getattr(cls,"__annotations__", {})

    @classmethod
    def describe(cls):
        print(f"Table: {cls.TableName()}")

        for field, typ in cls.fields().items():
            print(f" {field}: {typ.__name__}")

    def save(self):
        '''
            INSERT INTO {table_name} (columns)
        '''

        fields = self.__class__.fields()
        pk = self.primary_key()
        names = [n for n in fields ]
        values = [getattr(self,n, None) for n in names]
        placeholders = ", ".join("?" for _ in names)
        sql = f"INSERT INTO {self.__class__.TableName()} ({', '.join(names)}) VALUES ({placeholders})"


        conn = self.__class__.connection()
        cur = conn.cursor()
        cur.execute(sql, values)
        conn.commit()


    @classmethod
    def query(self):
        return QueryChainer(self)

    @classmethod
    def create_table(cls):
        fields = cls.fields()

        '''
            create table {table_name} (
                {name} {type} 
            )
        '''
        columns = []
        for name, typ in fields.items():
            sql_type = type_map.get(typ, 'TEXT')

            if isinstance(typ, PRIMARYKEY):            
                columns.append(f'{str(name)} {str(typ)}')
            elif isinstance(typ, FOREIGNKEY):
                cls._foreign_keys[typ.target_model] = name
                columns.append(f'{str(name)} {str(typ)}')
            else:
                columns.append(f'{str(name)} {type_map[typ]}')
        
        sql = f"CREATE TABLE IF NOT EXISTS {cls.TableName()} ({(','.join(columns))})"        
    
        conn = cls.connection()
        cur = conn.cursor()
        cur.execute(sql)
        conn.commit()



# YET TO implement to complex set of queries in here 
class QueryChainer:
    def __init__(self, model: BaseModel):
        self.model = model
        self._conditions = [] 
        self._kwargs = {}
        self._selected_fields = []
        self._order_by = []
        self._limit = None
        self._count = False


    def order_by(self, *fields):
        self._order_by = fields
        return self


    def count(self):
        self._count = True
        return self

    def limit(self, n):
        self._limit = n
        return self

    def select(self, *fields):
        """
        Accept Field objects like User.id, User.name
        """
        self._selected_fields = fields
        return self

    def filter(self,*args, **kwargs):
        
        for cond in args:
            if isinstance(cond, Condition):
                self._conditions.append(cond)
        

        for k, v in self._kwargs:
            self._kwargs[k] = v

        return self

    def _build_where_and_joins(self):
        joins = []
        clauses = []

        if self._conditions:
            if len(self._conditions) == 1:
                combined = self._conditions[0]
            else:
                combined = self._conditions[0]
                for cond in self._conditions[1:]:
                    combined = cond & combined

            where_clause = str(combined)


            for model in BaseModel.__subclasses__():
                if model.TableName() in where_clause and model != self.model:
                    joins.append(model)

            clauses.append(where_clause)

         # Process kwargs as normal
        for k, v in self._kwargs.items():
            val = f"'{v}'" if isinstance(v, str) else v
            clauses.append(f"{self.model.TableName()}.{k} = {val}")

        final_where = " AND ".join(clauses) if clauses else ""
        return final_where, joins


    def all(self):
        if self._count:
            field_list = "COUNT(*)"
        elif self._selected_fields:
            field_name = []
            for f in self._selected_fields:
                if f.model is not self.model:
                    raise ValueError("Invalid model in select class")

                field_name.append(f.full_name())
            
            field_list = ', '.join(field_name)
        else:
            field_list = f"{self.model.TableName()}.*"
        
        where_clause, joins = self._build_where_and_joins()
        sql = f"SELECT {field_list} FROM {self.model.TableName()}"

        for j in joins:
            fk_field = self.model._foreign_keys[j]
            sql += f" JOIN {j.TableName()} ON {self.model.TableName()}.{fk_field} = {j.TableName()}.id"

        if where_clause:
            sql += f" WHERE {where_clause}"

        if self._order_by:
            order_clause = ", ".join([f.full_name() for f in self._order_by])
            sql += f" ORDER BY {order_clause}"       

        if self._limit is not None:
            sql += f" LIMIT {self._limit}"


        print(sql)
        conn = self.model.connection()
        cur = conn.cursor()
        cur.execute(sql)
        rows = cur.fetchall()

        field_names = [f.name for f in self._selected_fields] if self._selected_fields else list(self.model.fields().keys())
        return [self.model(**dict(zip(field_names, row))) for row in rows]
    
    def first(self):
        results = self.all()
        return results[0] if results else None

File: phpormEngine/orm.php
<?php

namespace ORM;

use PDO;
use PDOException;
use ReflectionClass;
use ReflectionProperty;

// Type mapping
class TypeMap {
    public static $map = [
        'int' => 'INTEGER',
        'string' => 'TEXT',
        'float' => 'REAL',
        'bool' => 'INTEGER'
    ];
}

// Attribute classes
class PrimaryKey {
    public $baseType;
    
    public function __construct($type = 'int') {
        $this->baseType = $type;
    }
    
    public function __toString() {
        $sqlType = TypeMap::$map[$this->baseType] ?? 'TEXT';
        return "$sqlType PRIMARY KEY AUTOINCREMENT";
    }
}

class ForeignKey {
    public $targetModel;
    
    public function __construct($targetModel) {
        $this->targetModel = $targetModel;
    }
    
    public function __toString() {
        $tableName = call_user_func([$this->targetModel, 'tableName']);
        return "INTEGER REFERENCES {$tableName}(id)";
    }
}

class NotNull {
    public $baseType;
    
    public function __construct($type) {
        $this->baseType = $type;
    }
    
    public function __toString() {
        $sqlType = TypeMap::$map[$this->baseType] ?? 'TEXT';
        return "$sqlType NOT NULL";
    }
}

class Unique {
    public $baseType;
    
    public function __construct($type) {
        $this->baseType = $type;
    }
    
    public function __toString() {
        $sqlType = TypeMap::$map[$this->baseType] ?? 'TEXT';
        return "$sqlType UNIQUE";
    }
}

class DefaultValue {
    public $value;
    
    public function __construct($value) {
        $this->value = $value;
    }
    
    public function __toString() {
        if (is_string($this->value)) {
            return "DEFAULT '{$this->value}'";
        }
        return "DEFAULT {$this->value}";
    }
}

class Check {
    public $expr;
    
    public function __construct($expr) {
        $this->expr = $expr;
    }
    
    public function __toString() {
        return "CHECK({$this->expr})";
    }
}

// Condition class
class Condition {
    private $expr;
    
    public function __construct($expr) {
        $this->expr = $expr;
    }
    
    public function andCondition(Condition $other) {
        return new Condition("({$this->expr}) AND ({$other->expr})");
    }
    
    public function orCondition(Condition $other) {
        return new Condition("({$this->expr}) OR ({$other->expr})");
    }
    
    public function not() {
        return new Condition("NOT ({$this->expr})");
    }
    
    public function __toString() {
        return $this->expr;
    }
}

// Field class
class Field {
    private $name;
    private $model;
    
    public function __construct($name, $model = null) {
        $this->name = $name;
        $this->model = $model;
    }
    
    public function eq($value) {
        if ($value instanceof Field) {
            return new Condition("{$this->fullName()} = {$value->fullName()}");
        }
        $quoted = is_string($value) ? "'{$value}'" : $value;
        return new Condition("{$this->fullName()} = {$quoted}");
    }
    
    public function ne($value) {
        if ($value instanceof Field) {
            return new Condition("{$this->fullName()} != {$value->fullName()}");
        }
        $quoted = is_string($value) ? "'{$value}'" : $value;
        return new Condition("{$this->fullName()} != {$quoted}");
    }
    
    public function gt($value) {
        if ($value instanceof Field) {
            return new Condition("{$this->fullName()} > {$value->fullName()}");
        }
        $quoted = is_string($value) ? "'{$value}'" : $value;
        return new Condition("{$this->fullName()} > {$quoted}");
    }
    
    public function lt($value) {
        if ($value instanceof Field) {
            return new Condition("{$this->fullName()} < {$value->fullName()}");
        }
        $quoted = is_string($value) ? "'{$value}'" : $value;
        return new Condition("{$this->fullName()} < {$quoted}");
    }
    
    public function gte($value) {
        if ($value instanceof Field) {
            return new Condition("{$this->fullName()} >= {$value->fullName()}");
        }
        $quoted = is_string($value) ? "'{$value}'" : $value;
        return new Condition("{$this->fullName()} >= {$quoted}");
    }
    
    public function lte($value) {
        if ($value instanceof Field) {
            return new Condition("{$this->fullName()} <= {$value->fullName()}");
        }
        $quoted = is_string($value) ? "'{$value}'" : $value;
        return new Condition("{$this->fullName()} <= {$quoted}");
    }
    
    public function in($values) {
        $vals = array_map(function($v) {
            return is_string($v) ? "'{$v}'" : $v;
        }, $values);
        $valStr = implode(", ", $vals);
        return new Condition("{$this->fullName()} IN ({$valStr})");
    }
    
    public function like($pattern) {
        return new Condition("{$this->fullName()} LIKE '{$pattern}'");
    }
    
    public function fullName() {
        if ($this->model) {
            $tableName = call_user_func([$this->model, 'tableName']);
            return "{$tableName}.{$this->name}";
        }
        return $this->name;
    }
    
    public function getModel() {
        return $this->model;
    }
}

// Base Model class
abstract class BaseModel {
    protected static $dbPath = 'app.db';
    protected static $connection = null;
    public static $foreignKeys = [];
    private static $fieldCache = [];
    
    public function __construct(array $data = []) {
        $fields = static::fields();
        foreach ($fields as $name => $type) {
            $this->$name = $data[$name] ?? null;
        }
    }
    
    // Magic method to get Field objects as static properties
    public static function __callStatic($name, $arguments) {
        // Check if it's a valid field name
        if (array_key_exists($name, static::fields())) {
            // Cache the Field object for performance
            $cacheKey = static::class . '::' . $name;
            if (!isset(self::$fieldCache[$cacheKey])) {
                self::$fieldCache[$cacheKey] = new Field($name, static::class);
            }
            return self::$fieldCache[$cacheKey];
        }
        throw new \BadMethodCallException("Unknown field or method: {$name}");
    }
    
    public static function connection() {
        if (static::$connection === null) {
            try {
                static::$connection = new PDO('sqlite:' . static::$dbPath);
                static::$connection->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
            } catch (PDOException $e) {
                die("Connection failed: " . $e->getMessage());
            }
        }
        return static::$connection;
    }

    public function __toString() {
        $className = (new ReflectionClass($this))->getShortName();
        $fields = [];
        foreach (static::fields() as $name => $type) {
            $value = $this->$name ?? 'null';
            if (is_string($value)) {
                $value = "'{$value}'";
            }
            $fields[] = "{$name}={$value}";
        }
        return "{$className}(" . implode(", ", $fields) . ")";
    }
    
    public function __debugInfo() {
        $info = ['__class__' => get_class($this)];
        foreach (static::fields() as $name => $type) {
            $info[$name] = $this->$name ?? null;
        }
        return $info;
    }
    
    public static function tableName() {
        return strtolower((new ReflectionClass(static::class))->getShortName());
    }
    
    public static function primaryKey() {
        return 'id';
    }
    
    public static function fields() {
        // Override this in child classes
        return [];
    }
    
    public static function all() {
        $conn = static::connection();
        $tableName = static::tableName();
        $stmt = $conn->query("SELECT * FROM {$tableName}");
        $rows = $stmt->fetchAll(PDO::FETCH_ASSOC);
        
        $objects = [];
        foreach ($rows as $row) {
            $objects[] = new static($row);
        }
        return $objects;
    }
    
    public static function filter($conditions = null, array $kwargs = []) {
        $tableName = static::tableName();
        $joins = [];
        $whereClause = "";
        
        if ($conditions instanceof Condition) {
            $whereClause = (string)$conditions;
            
            // Auto-detect joins from condition
            foreach (static::$foreignKeys as $model => $fkField) {
                $joinTable = call_user_func([$model, 'tableName']);
                if (strpos($whereClause, $joinTable) !== false) {
                    $joins[] = $model;
                }
            }
        } else {
            $conditionsArr = [];
            foreach ($kwargs as $k => $v) {
                $val = is_string($v) ? "'{$v}'" : $v;
                $conditionsArr[] = "{$tableName}.{$k} = {$val}";
            }
            $whereClause = implode(" AND ", $conditionsArr);
        }
        
        $sql = "SELECT {$tableName}.* FROM {$tableName}";
        
        // Add JOINs
        foreach ($joins as $joinModel) {
            $joinTable = call_user_func([$joinModel, 'tableName']);
            $fkField = static::$foreignKeys[$joinModel];
            $sql .= " JOIN {$joinTable} ON {$tableName}.{$fkField} = {$joinTable}.id";
        }
        
        if ($whereClause) {
            $sql .= " WHERE {$whereClause}";
        }
        
        $conn = static::connection();
        $stmt = $conn->query($sql);
        $rows = $stmt->fetchAll(PDO::FETCH_ASSOC);
        
        $objects = [];
        foreach ($rows as $row) {
            $objects[] = new static($row);
        }
        return $objects;
    }
    
    public static function update(array $where = [], array $data = []) {
        if (empty($data)) {
            throw new \Exception("No fields to update");
        }
        
        $tableName = static::tableName();
        $setClause = [];
        $setValues = [];
        
        foreach ($data as $k => $v) {
            $setClause[] = "{$k} = ?";
            $setValues[] = $v;
        }
        
        $whereClause = "";
        $whereValues = [];
        if (!empty($where)) {
            $whereParts = [];
            foreach ($where as $k => $v) {
                $whereParts[] = "{$k} = ?";
                $whereValues[] = $v;
            }
            $whereClause = " WHERE " . implode(" AND ", $whereParts);
        }
        
        $sql = "UPDATE {$tableName} SET " . implode(", ", $setClause) . $whereClause;
        
        $conn = static::connection();
        $stmt = $conn->prepare($sql);
        $stmt->execute(array_merge($setValues, $whereValues));
    }
    
    public function save() {
        $fields = static::fields();
        $names = array_keys($fields);
        $values = [];
        
        foreach ($names as $name) {
            $values[] = $this->$name ?? null;
        }
        
        $placeholders = implode(", ", array_fill(0, count($names), "?"));
        $tableName = static::tableName();
        $sql = "INSERT INTO {$tableName} (" . implode(", ", $names) . ") VALUES ({$placeholders})";
        
        $conn = static::connection();
        $stmt = $conn->prepare($sql);
        $stmt->execute($values);
        
        $pk = static::primaryKey();
        if ($pk) {
            $this->$pk = $conn->lastInsertId();
        }
    }
    
    public static function query() {
        return new QueryChainer(static::class);
    }
    
    public static function createTable() {
        $fields = static::fields();
        $tableName = static::tableName();
        $columns = [];
        
        foreach ($fields as $name => $type) {
            if ($type instanceof PrimaryKey) {
                $columns[] = "{$name} {$type}";
            } elseif ($type instanceof ForeignKey) {
                static::$foreignKeys[$type->targetModel] = $name;
                $columns[] = "{$name} {$type}";
            } elseif (isset(TypeMap::$map[$type])) {
                $columns[] = "{$name} " . TypeMap::$map[$type];
            } else {
                $columns[] = "{$name} {$type}";
            }
        }
        
        $sql = "CREATE TABLE IF NOT EXISTS {$tableName} (" . implode(", ", $columns) . ")";
        
        $conn = static::connection();
        $conn->exec($sql);
    }
    
    public static function describe() {
        echo "Table: " . static::tableName() . "\n";
        foreach (static::fields() as $field => $type) {
            $typeName = is_object($type) ? get_class($type) : $type;
            echo " {$field}: {$typeName}\n";
        }
    }
}

// Query Chainer class
class QueryChainer {
    private $model;
    private $conditions = [];
    private $kwargs = [];
    private $selectedFields = [];
    private $orderBy = [];
    private $groupBy = [];
    private $limit = null;
    private $count = false;
    private $typeFiller = null;
    
    public function __construct($model) {
        $this->model = $model;
    }
    
    public function orderBy(...$fields) {
        $this->orderBy = $fields;
        return $this;
    }
    
    public function groupBy(...$fields) {
        $this->groupBy = $fields;
        return $this;
    }
    
    public function count() {
        $this->count = true;
        return $this;
    }
    
    public function limit($n) {
        $this->limit = $n;
        return $this;
    }
    
    public function select(...$fields) {
        $this->selectedFields = $fields;
        return $this;
    }
    
    public function filter($conditions = null, array $kwargs = []) {
        if ($conditions instanceof Condition) {
            $this->conditions[] = $conditions;
        }
        $this->kwargs = array_merge($this->kwargs, $kwargs);
        return $this;
    }
    
    public function fillType($model) {
        $this->typeFiller = $model;
        return $this;
    }
    
    private function buildWhereAndJoins() {
        $joins = [];
        $clauses = [];
        
        if (!empty($this->conditions)) {
            $combined = $this->conditions[0];
            for ($i = 1; $i < count($this->conditions); $i++) {
                $combined = $combined->andCondition($this->conditions[$i]);
            }
            
            $whereClause = (string)$combined;
            
            // Collect all foreign key models from the main model
            $foreignKeys = $this->model::$foreignKeys ?? [];
            // var_dump($this->model::$foreignKeys);

            // Also check selected fields for foreign models
            $referencedModels = [];
            foreach ($this->selectedFields as $field) {
                if ($field instanceof Field) {
                    $fieldModel = $field->getModel();
                    if ($fieldModel && $fieldModel !== $this->model) {
                        $referencedModels[$fieldModel] = true;
                    }
                }
            }
            
            // Auto-detect joins from WHERE clause - check ALL models
            foreach ($foreignKeys as $modelClass => $fkField) {
                $tableName = call_user_func([$modelClass, 'tableName']);
                if (strpos($whereClause, $tableName) !== false || isset($referencedModels[$modelClass])) {
                    $joins[$modelClass] = $fkField;
                }
            }
            
            $clauses[] = $whereClause;
        }
        
        foreach ($this->kwargs as $k => $v) {
            $val = is_string($v) ? "'{$v}'" : $v;
            $tableName = call_user_func([$this->model, 'tableName']);
            $clauses[] = "{$tableName}.{$k} = {$val}";
        }
        
        $finalWhere = !empty($clauses) ? implode(" AND ", $clauses) : "";
        return [$finalWhere, $joins];
    }
    
    public function toSql() {
        $tableName = call_user_func([$this->model, 'tableName']);
        
        if ($this->count) {
            if (!empty($this->groupBy)) {
                $groupFields = array_map(function($f) {
                    return $f->fullName();
                }, $this->groupBy);
                $fieldList = implode(", ", $groupFields) . ", COUNT(*)";
            } else {
                $fieldList = "COUNT(*)";
            }
        } elseif (!empty($this->selectedFields)) {
            $fieldNames = array_map(function($f) {
                return $f->fullName();
            }, $this->selectedFields);
            $fieldList = implode(", ", $fieldNames);
        } else {
            $fieldList = "{$tableName}.*";
        }
        
        list($whereClause, $joins) = $this->buildWhereAndJoins();
        $sql = "SELECT {$fieldList} FROM {$tableName}";
        
        // Add JOINs - now $joins is an associative array [Model => fk_field]
        foreach ($joins as $joinModel => $fkField) {
            $joinTable = call_user_func([$joinModel, 'tableName']);
            $sql .= " JOIN {$joinTable} ON {$tableName}.{$fkField} = {$joinTable}.id";
        }
        
        if ($whereClause) {
            $sql .= " WHERE {$whereClause}";
        }
        
        if (!empty($this->groupBy)) {
            $groupFields = array_map(function($f) {
                return $f->fullName();
            }, $this->groupBy);
            $sql .= " GROUP BY " . implode(", ", $groupFields);
        }
        
        if (!empty($this->orderBy)) {
            $orderFields = array_map(function($f) {
                return $f->fullName();
            }, $this->orderBy);
            $sql .= " ORDER BY " . implode(", ", $orderFields);
        }
        
        if ($this->limit !== null) {
            $sql .= " LIMIT {$this->limit}";
        }
        
        return $sql;
    }
    
    public function all($jsonEnable = false) {
        $sql = $this->toSql();
        $conn = call_user_func([$this->model, 'connection']);
        $stmt = $conn->query($sql);
        $rows = $stmt->fetchAll(PDO::FETCH_ASSOC);
        
        if ($jsonEnable) {
            return json_encode($rows, JSON_PRETTY_PRINT);
        }
        
        $objects = [];
        foreach ($rows as $row) {
            if ($this->typeFiller) {
                $objects[] = new $this->typeFiller($row);
            } else {
                $objects[] = new $this->model($row);
            }
        }
        return $objects;
    }
    
    public function first() {
        $results = $this->all();
        return !empty($results) ? $results[0] : null;
    }
}

// Example Models
class User extends BaseModel {
    public $id;
    public $name;
    public $email;
    public $age;
    public $country;
    
    public static function fields() {
        return [
            'id' => new PrimaryKey('int'),
            'name' => 'string',
            'email' => 'string',
            'age' => 'int',
            'country' => 'string'
        ];
    }
}

class Post extends BaseModel {
    public $id;
    public $title;
    public $content;
    public $user_id;
    public $views;
    
    public static function fields() {
        return [
            'id' => new PrimaryKey('int'),
            'title' => 'string',
            'content' => 'string',
            'user_id' => new ForeignKey(User::class),
            'views' => 'int'
        ];
    }
}

class Author extends BaseModel {
    public $id;
    public $name;
    public $bio;
    
    public static function fields() {
        return [
            'id' => new PrimaryKey('int'),
            'name' => 'string',
            'bio' => 'string'
        ];
    }
}

class Publisher extends BaseModel {
    public $id;
    public $name;
    public $address;
    
    public static function fields() {
        return [
            'id' => new PrimaryKey('int'),
            'name' => 'string',
            'address' => 'string'
        ];
    }
}

class Book extends BaseModel {
    public $id;
    public $title;
    public $author_id;
    public $publisher_id;
    public $pages;
    
    public static function fields() {
        return [
            'id' => new PrimaryKey('int'),
            'title' => 'string',
            'author_id' => new ForeignKey(Author::class),
            'publisher_id' => new ForeignKey(Publisher::class),
            'pages' => 'int'
        ];
    }
}

// ============= TEST SCRIPT =============
echo "=== Creating Tables ===\n";
User::createTable();
Post::createTable();
Author::createTable();
Publisher::createTable();
Book::createTable();


// $q = Book::query()
//     ->select(Book::id(), Publisher::name())->filter(Book::id()->eq(2)->andCondition(Publisher::name()->eq("Meow")))->toSql();

// echo $q . "\n";


$q = Book::query()->select(Book::id())->filter(Book::title()->like("%Me"));

echo $q->toSql() . "\n";

// var_dump(User::$foreignKeys);

?>